File: src/domain/nutrition.ts
import type { Nutrition } from "../types/models";

export function zeroNutrition(): Nutrition {
  return { kcal: 0, proteinG: 0, fatG: 0, carbsG: 0, fiberG: 0, sugarG: 0, saltG: 0 };
}

export function nutr(
  kcal: number,
  proteinG: number,
  fatG: number,
  carbsG: number,
  extra: Partial<Nutrition> = {},
): Nutrition {
  return {
    kcal,
    proteinG,
    fatG,
    carbsG,
    fiberG: extra.fiberG ?? 0,
    sugarG: extra.sugarG ?? 0,
    saltG: extra.saltG ?? 0,
  };
}

export function scaleNutrition(n: Nutrition, factor: number): Nutrition {
  return {
    kcal: n.kcal * factor,
    proteinG: n.proteinG * factor,
    fatG: n.fatG * factor,
    carbsG: n.carbsG * factor,
    fiberG: n.fiberG * factor,
    sugarG: n.sugarG * factor,
    saltG: n.saltG * factor,
  };
}

export function addNutrition(a: Nutrition, b: Nutrition): Nutrition {
  return {
    kcal: a.kcal + b.kcal,
    proteinG: a.proteinG + b.proteinG,
    fatG: a.fatG + b.fatG,
    carbsG: a.carbsG + b.carbsG,
    fiberG: a.fiberG + b.fiberG,
    sugarG: a.sugarG + b.sugarG,
    saltG: a.saltG + b.saltG,
  };
}

export function nutritionForGrams(per100: Nutrition, grams: number): Nutrition {
  return scaleNutrition(per100, grams / 100);
}

export function roundNutrition(n: Nutrition): Nutrition {
  return {
    kcal: Math.round(n.kcal),
    proteinG: Math.round(n.proteinG * 10) / 10,
    fatG: Math.round(n.fatG * 10) / 10,
    carbsG: Math.round(n.carbsG * 10) / 10,
    fiberG: Math.round(n.fiberG * 10) / 10,
    sugarG: Math.round(n.sugarG * 10) / 10,
    saltG: Math.round(n.saltG * 100) / 100,
  };
}

export function formatNutrition(n: Nutrition): string {
  const r = roundNutrition(n);
  return `${r.kcal} ккал · Б ${r.proteinG} · Ж ${r.fatG} · У ${r.carbsG}`;
}

export const ADULT_NORM = { kcal: 2200, proteinG: 75, fatG: 70, carbsG: 280 };
export const SCHOOL_NORM = { kcal: 1900, proteinG: 70, fatG: 65, carbsG: 260 };

export function coverage(actual: Nutrition, norm = ADULT_NORM) {
  const pct = (part: number, whole: number) => (whole === 0 ? 0 : (part / whole) * 100);
  return {
    kcalPct: pct(actual.kcal, norm.kcal),
    proteinPct: pct(actual.proteinG, norm.proteinG),
    fatPct: pct(actual.fatG, norm.fatG),
    carbsPct: pct(actual.carbsG, norm.carbsG),
  };
}

File: rust/src/nutrition.rs
use serde::{Deserialize, Serialize};

/// Пищевая ценность на фиксированную базу (обычно 100 г).
#[derive(Debug, Clone, Copy, Default, PartialEq, Serialize, Deserialize)]
pub struct Nutrition {
    pub kcal: f64,
    pub protein_g: f64,
    pub fat_g: f64,
    pub carbs_g: f64,
    #[serde(default)]
    pub fiber_g: f64,
    #[serde(default)]
    pub sugar_g: f64,
    #[serde(default)]
    pub salt_g: f64,
}

impl Nutrition {
    pub fn zero() -> Self {
        Self::default()
    }

    pub fn scale(self, factor: f64) -> Self {
        Self {
            kcal: self.kcal * factor,
            protein_g: self.protein_g * factor,
            fat_g: self.fat_g * factor,
            carbs_g: self.carbs_g * factor,
            fiber_g: self.fiber_g * factor,
            sugar_g: self.sugar_g * factor,
            salt_g: self.salt_g * factor,
        }
    }

    pub fn add(self, other: Self) -> Self {
        Self {
            kcal: self.kcal + other.kcal,
            protein_g: self.protein_g + other.protein_g,
            fat_g: self.fat_g + other.fat_g,
            carbs_g: self.carbs_g + other.carbs_g,
            fiber_g: self.fiber_g + other.fiber_g,
            sugar_g: self.sugar_g + other.sugar_g,
            salt_g: self.salt_g + other.salt_g,
        }
    }

    /// Пересчёт с 100 г на фактический вес съедобной части.
    pub fn for_grams(self, grams: f64) -> Self {
        self.scale(grams / 100.0)
    }

    pub fn calories_from_macros(self) -> f64 {
        self.protein_g * 4.0 + self.carbs_g * 4.0 + self.fat_g * 9.0
    }

    pub fn rounded(self) -> Self {
        Self {
            kcal: self.kcal.round(),
            protein_g: round1(self.protein_g),
            fat_g: round1(self.fat_g),
            carbs_g: round1(self.carbs_g),
            fiber_g: round1(self.fiber_g),
            sugar_g: round1(self.sugar_g),
            salt_g: round2(self.salt_g),
        }
    }
}

fn round1(v: f64) -> f64 {
    (v * 10.0).round() / 10.0
}

fn round2(v: f64) -> f64 {
    (v * 100.0).round() / 100.0
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct DailyNutritionNorm {
    pub kcal: f64,
    pub protein_g: f64,
    pub fat_g: f64,
    pub carbs_g: f64,
}

impl DailyNutritionNorm {
    pub fn adult_mixed() -> Self {
        Self {
            kcal: 2200.0,
            protein_g: 75.0,
            fat_g: 70.0,
            carbs_g: 280.0,
        }
    }

    pub fn school() -> Self {
        Self {
            kcal: 1900.0,
            protein_g: 70.0,
            fat_g: 65.0,
            carbs_g: 260.0,
        }
    }

    pub fn coverage(self, actual: Nutrition) -> NutritionCoverage {
        NutritionCoverage {
            kcal_pct: pct(actual.kcal, self.kcal),
            protein_pct: pct(actual.protein_g, self.protein_g),
            fat_pct: pct(actual.fat_g, self.fat_g),
            carbs_pct: pct(actual.carbs_g, self.carbs_g),
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct NutritionCoverage {
    pub kcal_pct: f64,
    pub protein_pct: f64,
    pub fat_pct: f64,
    pub carbs_pct: f64,
}

fn pct(part: f64, norm: f64) -> f64 {
    if norm.abs() < 1e-9 {
        0.0
    } else {
        part / norm * 100.0
    }
}
